### PR TITLE
Fix `<=` and `>` comparisons when compared against prerelease versions

### LINKED ISSUE #50909: typesVersions should not have `4.9.0-beta <=4.8`
currently, a typesVersion entry of `"<=4.8"` will be satisfied by 4.9.0-beta.

That's because our semver code treats `"4.8"` as the range ">=4.8.0 and <4.9.0". It should actually be ">=4.8.0 and <4.8.Infinity". It should not include 4.9.0-*xxx*.

### PR DESCRIPTION
**Pull request by rbuckton**
This fixes an issue with semver comparisons where `<=` and `>` comparison against a partial version (i.e., `<=4.8` or `<=4.8.x`) accidentally include prerelease versions of the next higher patch or minor version. this is because `<=4.8` is simplified to `<4.9.0`, but should have been simplified to `<4.9.0-0` because prerelease versions are considered lower than their non-prerelease counterparts. In the same vein, `>` comparisons like `>4.8` were incorrectly simplified to `>=4.9.0`, but should have been simplified to `>=4.9.0-0`.

Fixes #50909

### CODE REVIEW

File: src/testRunner/unittests/semver.ts [hunk 1]
```diff
@@ -167,6 +168,7 @@ namespace ts {
                 ["<=2", "2.0.0"],
                 ["<=2.1", "2.1.0"],
                 ["<=2.0.1", "2.0.1"],
+                ["<=4.8", "4.9.0-beta", false],
```
Comment (DanielRosenwasser) [inline src/testRunner/unittests/semver.ts:171]:
```suggestion
                ["<=4.8", "4.9.0-beta", false],
                [">=4.9", "4.9.0-beta", true],
```

**Comment by fatcerberus** [comment]
`>4.8` is an interesting case.  As a human, I want to read that as "4.9 or higher", because it just seems intuitive that the greater-than comparison would apply to the least significant digit in the stated version number (points in version numbers are not decimal points).  Normalizing `4.8` to `4.8.0` isn't something that would ever occur to me in the context of a comparison.

File: src/testRunner/unittests/semver.ts (hunk 1 above)
Comment (rbuckton) [reply src/testRunner/unittests/semver.ts:171]: `>=4.9` [matches `4.9.0-beta`](https://jubianchi.github.io/semver-check/#/%3E%3D4.9.0/4.9.0-beta), but `>4.9` [does not](https://jubianchi.github.io/semver-check/#/%3E4.9/4.9.0-beta).

semver is weird

**Comment by rbuckton** [comment]
Sorry, that's a typo. `>4.8` is satisfied by `4.9.0-0`.

**Comment by fatcerberus** [comment]
Okay, so it does align with my intuition.  Good to know.

Semver is definitely weird.  It’s not even a total order.  `1.0.0-beta` is less than `1.0.0` but I’m pretty sure all suffixes are considered equal to each other so there’s no suffix you can use as a “least” value…

**Comment by rbuckton** [comment]
Suffixes are ordered. `-0` is less than `-1`, and any number segment is less than any segment containing an alpha character, thus `-0` is less than `-alpha`, and `-alpha` is less than `-beta`. The _least_ prerelease suffix is `-0`.

**Comment by fatcerberus** [comment]
Huh, TIL.  I would have thought if the suffixes were ordered at all it would always be a string comparison.  Most people don’t use numbers after the dash in my experience.

**Comment by rbuckton** [comment]
https://semver.org/#spec-item-11, step 4. Though `node-semver` only compares prerelease versions when the major, minor, and patch versions are identical unless you pass `includePrerelease` (https://github.com/npm/node-semver#prerelease-tags), making everything a bit inconsistent.

**Comment by jakebailey** [comment]
> Most people don’t use numbers after the dash in my experience.

I actually think it's more common that way, e.g. `-beta.1`, `-beta.2`, `-rc.5`, etc, as prerelease versions sort by splitting on `.`, so as long as you sort the non-number portions how you want (alpha comes before beta comes before rc), thinks work. For example, [eslint prereleases](https://www.npmjs.com/package/eslint/v/8.0.0-beta.2).

> Though node-semver only compares prerelease versions when the major, minor, and patch versions are identical unless you pass includePrerelease ([npm/node-semver#prerelease-tags](https://github.com/npm/node-semver#prerelease-tags)), making everything a bit inconsistent.

Just to confirm, this PR makes our semver behave like node-semve when `includePrerelease` is enabled, correct? I think we discussed that having that false like semver does by default is useful for package managers but not so much us who are just trying to correctly semantically order things.

**Comment by rbuckton** [comment]
IIRC, our behavior has always been to match `includePrerelease: true` since we're not a package manager. This has just always been a bug.

**Comment by rbuckton** [comment]
> IIRC, our behavior has always been to match `includePrerelease: true` since we're not a package manager. This has just always been a bug.

Also since we mostly use this to test our own versions, which don't strictly follow semantic versioning...

**Comment by jakebailey** [comment]
Yeah, it's a little unfortunate that we use `-dev` as the nightly suffix when technically that sorts after `beta`. I've had confusing issues installing the beta thanks to that 🙁.

**Comment by rbuckton** [comment]
I found another bug, I'm writing up some more comprehensive tests to make sure I've got everything else covered.

**Comment by rbuckton** [comment]
This should be ready for another look @sandersn, @DanielRosenwasser

File: package-lock.json [hunk 2]
```diff
@@ -9953,7 +9953,7 @@
             "integrity": "sha512-m62ShEObQ39CfralilEQRjH6oAMtNCV1xJyEx5LpRYUVN+EviphDgUc/F3hnYbADmkiNs67Y+3ylmlG7Lnu+FA==",
             "dev": true,
             "requires": {
-                "es5-ext": "0.10.53",
+                "es5-ext": "^0.10.50",
```
Comment (jakebailey) [inline package-lock.json:9956]: Not sure what happened here, but I don't think these package-lock changes were intentional.

Comment (rbuckton) [reply package-lock.json:9956]: Oi, nope. I just did an `npm i` because `gulp lint` was complaining locally. I'll fix.

Comment (jakebailey) [reply package-lock.json:9956]: If you aren't already, I'd recommend enabling corepack, that way you run the version of npm declared in the package.json. Though, if a newer npm is doing this, we can totally bump that and relock.

Comment (jakebailey) [reply package-lock.json:9956]: (Or the other way around, maybe we are relocking with a different version of npm; the GHA workflow doesn't enable corepack, something I forgot to do...)

File: src/testRunner/unittests/semver.ts [hunk 3]
```diff
@@ -120,129 +98,826 @@ namespace ts {
...
+                ["*", "1.0.1-0", true],
+                ["*", "1.0.0", true],
+                ["*", "1.0.0-0", true],
+                ["*", "0.0.0", true],
+                ["*", "0.0.0-0", true],
+
+                // wildcard minor
                 ["1", "2.0.0", false],
-                ["1.0", "1.0.0"],
+                ["1", "2.0.0-0", false], // failing
```
Comment (sandersn) [inline src/testRunner/unittests/semver.ts:212]: what does `// failing` mean here?

**Review by sandersn — APPROVED** [review]

File: src/testRunner/unittests/semver.ts (hunk 3 above)
Comment (rbuckton) [reply src/testRunner/unittests/semver.ts]: Nothing important. I've removed it, thanks.

**Pull request merged**